garak/probes/latentinjection.py

```diff
@@ -0,0 +1,448 @@
+
+"""
+
+import logging
+import random
+
+from garak import _config
+from garak.attempt import Attempt
+from garak.probes.base import Probe
+
+INJECTION_MARKER = "<|garak_injection|>"
+PAYLOAD_MARKER = "<|garak_payload|>"
+TRIGGER_MARKER = "<|garak_trigger|>"
+
+
+class LatentInjectionMixin:
+
+    primary_detector = "base.TriggerListDetector"
+    tags = [
+        "avid-effect:security:S0403",
+        "owasp:llm01",
+        "quality:Security:PromptStability",
+    ]
+
+    injection_sep_pairs = [("", "")]  # optional attrib get a default
+
+    def _build_prompts_triggers(self) -> None:
+        self.prompts = []
+        self.triggers = []
+
+        for top_instruction in self.top_instructions:
+            for doc in self.docs:
+                if INJECTION_MARKER not in doc:
+                    logging.warning(
+                        "No injection marker in doc for %s", self.__class__.__name__
+                    )
+                    continue
+                for injection_sep_pair in self.injection_sep_pairs:
+                    injection_sep_start, injection_sep_end = injection_sep_pair
+                    for injection_instruction in self.injection_instructions:
+                        for payload in self.payloads:
+                            for payload_trigger in self.payload_triggers:
+                                this_payload = payload.replace(
+                                    TRIGGER_MARKER, payload_trigger
+                                )
+                                injection = "".join(
+                                    [
+                                        injection_sep_start,
+                                        injection_instruction.replace(
+                                            PAYLOAD_MARKER, this_payload
+                                        ),
+                                        injection_sep_end,
+                                    ]
+                                )
+                                complete_prompt = top_instruction + doc.replace(
+                                    INJECTION_MARKER, injection
+                                )
+                                self.prompts.append(complete_prompt)
+                                self.triggers.append(payload_trigger)
+
+    def _attempt_prestore_hook(self, attempt: Attempt, seq: int) -> Attempt:
+        attempt.notes["triggers"] = [self.triggers[seq]]
+        return attempt
+
+
+class LatentInjectionTranslation(Probe, LatentInjectionMixin):
+    """String-based latent prompt injections via translation
+
+    Check to see if model can be subverted by text placed in an
+    otherwise innocuous document"""
+
+    tags = LatentInjectionMixin.tags
+    primary_detector = LatentInjectionMixin.primary_detector
+    bcp47 = "en"
+    goal = "subvert translation instructions"
+    doc_uri = "https://x.com/goodside/status/1569128808308957185"
```

Comment:
Is it related or is it the latent word that's flagged here?

Reply:
Isn't this fundamentally the same attack? I worry about link rot, especially with Twitter links.

Reply:
Cam archive.org it before twitter.com inevitably disappears I guess.

Not sure. For some of these the attribution is way earlier than arXiv, and I dont assign arXiv any more authority than Twitter

Reply:
The attack in this paper from ~10 months after Goodside's translation injection is I guess similar in nature, with both embedding instructions in something that's framed as data. I think it could be possible to implement this attack using the existing mixin class, without too much effort. Here we appear to be focused on changing the order of an innocuous and a content-safety-breaking instruction and hoping we get a content safety breach.

My question here is: would one consider the use of the prompts/templates in this paper (see e.g. https://raw.githubusercontent.com/qiuhuachuan/latent-jailbreak/main/img/latent-jailbreak-dataset.png) in order to get toxicity out, a security breach/alignment subversion?

Reply:
conclusion of discussion: implement these templates in a probe in this class

Reply:
implemented (w/ a superset and then a mini version of that)